### PR TITLE
Implement block restriction of ElementLoopObject in rdg module

### PR DESCRIPTION
**Pull request by lindsayad**
- Make `ElementLoopObject` inherit from `BlockRestrictable`
- In `onElement` and `onInternalSide` methods, only call respective compute functions if elements reside on blocks for which the `ElementLoopObject` exists
- Modify `AEFVSlopeLimiting1D` so that it doesn't try to get values on a neighboring element on a subdomain on which the base `ElementLoopObject` is not defined
- added block restricted test; only slightly modified from base 1d_advection_tests

Refs #8357

### CODE REVIEW

**Comment by lindsayad** [comment]
@yidongxiainl @andrsd I'm working on a project for which I need block restriction ASAP, so I've done some coding work myself that makes it work. You guys can accept or reject this as far as incorporation into the framework as concerned.

**Comment by mangerij** [comment]
How exactly are you going to get this to work in parallel if the neighbor
is outside of the block? How do the processors communicate information
about the neighbor?

This is an issue that we ran into in the last few months

**Comment by lindsayad** [comment]
Is there an issue ticket or example that you can point me to so I can learn more about the problems you've had? What systems (e.g. DGKernel, InterfaceKernel) have you had the issue with? That could inform my design choices.

**Comment by yidongxiainl** [comment]
@lindsayad This is the PR <https://github.com/idaholab/moose/pull/8344>. This is my branch <https://github.com/yidongxiainl/moose/tree/xiay/issue8228>. I will also send you a technical report.

**Comment by andrsd** [comment]
@mangerij If I understand this PR, when we are on the boundary of a subdomain restricted variable, he uses a know value, he is not going to use the values from a neighbor. Right @lindsayad ?

For general DG, if a neighbor is owned by a another CPU, its values should still be ghosted and therefore accessible to the active CPU. For rDG it is a little more complicated, because we are reconstructing gradients and computing slope limiters, but the values from interfaces are send to neighboring CPUs. Not sure if that answers your question...

**Comment by lindsayad** [comment]
On 01/23/2017 09:43 AM, David Andrs wrote:
>
> @mangerij <https://github.com/mangerij> If I understand this PR, when 
> we are on the boundary of a subdomain restricted variable, he uses a 
> know value, he is not going to use the values from a neighbor. Right 
> @lindsayad <https://github.com/lindsayad> ?
>
Right, for a variable and user_object restricted to a subdomain, the 
behavior on the subdomain boundary should be the same as if we were on 
the boundary of a single domain problem.

I will say that this block restriction doesn't make much sense for 1D 
problems, which are the only problems that the rDG module currently has 
limiters for. Where it starts to make sense is in multi-dimensional 
problems where you have material (and variable boundaries) perpendicular 
to the direction of advective flow.

**Comment by andrsd** [comment]
@lindsayad What do you think about changing the logic in `ElementLoopUserObject::execute()` in a way that when on true internal edge we call `onInternalSide`, when on a true boundary or a subdomain boundary we call `onBoundary`?

**Comment by andrsd** [comment]
Also, similarly moving the if statement from `ElementLoopUserObject::onElement()` to `execute()`...

**Comment by lindsayad** [comment]
Isn't the logic currently doing that with the exception that if we're on 
a subdomain boundary `onInterface` is getting called? That code looks to 
be copied verbatim from `ThreadedElementLoopBase`. I probably don't 
understand what you're getting at.

Where I'm a little troubled is that for the current rDG limiters, all 
the logic gets executed through computeElement(). On one hand this makes 
sense because the limiter is being applied to the current element. On 
the other hand we depend heavily on neighbor information when we're 
internal, so it almost feels like logic should be executed through the 
computeInternalSide() method, and then through computeBoundary() when 
we're on a domain (or subdomain) boundary. Maybe that's kind of what 
you're thinking when talking about modifying the logic in `execute()`? 
But then again there's currently no way that the `ElementLoopObject` 
knows what it's ultimately being used for since it's the base class and 
all it really has is geometric information (elem, etc).

**Comment by lindsayad** [comment]
This makes good sense.

**Comment by andrsd** [comment]
> Isn't the logic currently doing that with the exception that if we're on a subdomain boundary `onInterface` is getting called? That code looks to be copied verbatim from `ThreadedElementLoopBase`. I probably don't understand what you're getting at.

What I was thinking was that `onBoundary` should be called when we want to apply a "boundary condition". Either when we are on the physical boundary of the domain or a boundary of a block restricted variable. But, it is probably not a good idea, since we have `onInterface` for the latter part as you mentioned. Still, the added if-statement in `onInternalSide` should probably go into `execute()` similarly to the change in `onElement`.


> Where I'm a little troubled is that for the current rDG limiters, all the logic gets executed through computeElement().

If you look at some of our slope limiting code, it typically precomputes some quantity on an element  using all neighbors and then it uses it to compute the limited slope. If using `onInternalSide()`, we would need to add some other call after we are done looping over edges. This is possible, but the way we did it is simpler, since the routine is not split into 2 parts.

I am not saying this is the final design, it is what we currently have and seems to work.

File: modules/rdg/src/userobjects/ElementLoopUserObject.C [hunk 1]
```diff
@@ -136,8 +143,8 @@ ElementLoopUserObject::subdomainChanged()
 void
 ElementLoopUserObject::onElement(const Elem * elem)
 {
-  _current_elem = elem;
-  computeElement();
+    _current_elem = elem;
+    computeElement();
```
Comment (andrsd) [inline modules/rdg/src/userobjects/ElementLoopUserObject.C]: Leftover indentation here. Should be 2 spaces...

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by lindsayad** [comment]
> If you look at some of our slope limiting code, it typically precomputes some quantity on an element using all neighbors and then it uses it to compute the limited slope. If using onInternalSide(), we would need to add some other call after we are done looping over edges. This is possible, but the way we did it is simpler, since the routine is not split into 2 parts.

No, I agree with what you guys did.

**Review by andrsd — APPROVED** [review]

**Comment by andrsd** [comment]
Thanks for the PR @lindsayad

**Comment by lindsayad** [comment]
You're welcome. Thanks for the design discussion. I look forward to 
further rDG development!

**Pull request merged**